tests/unit/test_versioning.py

```diff
@@ -0,0 +1,181 @@
+"""
+ This file is part of nucypher.
+
+ nucypher is free software: you can redistribute it and/or modify
+ it under the terms of the GNU Affero General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ nucypher is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU Affero General Public License for more details.
+
+ You should have received a copy of the GNU Affero General Public License
+ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+
+from typing import Tuple, Any, Type
+
+import pytest
+
+from nucypher.utilities.versioning import Versioned
+
+
+def _check_valid_version_tuple(version: Any, cls: Type):
+    if not isinstance(version, tuple):
+        pytest.fail(f"Old version handlers keys for {cls.__name__} must be a tuple")
+    if not len(version) == Versioned._PARTS:
+        pytest.fail(f"Old version handlers keys for {cls.__name__} must be a {str(Versioned._BRAND_SIZE)}-tuple")
+    if not all(isinstance(part, int) for part in version):
+        pytest.fail(f"Old version handlers version parts {cls.__name__} must be integers")
+
+
+class A(Versioned):
+
+    def __init__(self, x):
+        self.x = x
+
+    @classmethod
+    def _brand(cls):
+        return b"AA"
+
+    @classmethod
+    def _version(cls) -> Tuple[int, int]:
+        return 2, 1
+
+    def _payload(self) -> bytes:
+        return bytes(self.x)
+
+    @classmethod
+    def _old_version_handlers(cls):
+        return {
+            (1, 0): cls._from_bytes_v1_0,
+            (2, 0): cls._from_bytes_v2_0,
+        }
+
+    @classmethod
+    def _from_bytes_v1_0(cls, data):
+        return cls(int(data))  # we used to keep it in decimal representation
+
+    @classmethod
+    def _from_bytes_v2_0(cls, data):
+        return cls(int(data, 16))  # then we switched to the hexadecimal
+
+    @classmethod
+    def _from_bytes_current(cls, data):
+        return cls(str(int(data, 16)))  # but now we use a string representation for some reason
+
+
+def test_unique_branding():
+    brands = tuple(v._brand() for v in Versioned.__subclasses__())
+    if len(brands) != len(set(brands)):
+        pytest.fail(f"Duplicated_brand {Versioned._brand()}.")
```

Comment:
Doesn't this test just raise a `NotImplementedError` if the test fails - is this intended? Perhaps you can instead show the duplicate - not the most efficient but 1) it is a unit test, and 2) only happens in the failure case)

```suggestion
    brands = tuple(v._brand() for v in Versioned.__subclasses__())
    brands_set = set(brands)
    if len(brands) != len(brands_set):
        duplicate_brands = list(brands)
        for brand in brands_set:
            duplicate_brands.remove(brand)
        pytest.fail(f"Duplicated brand(s) {duplicate_brands}.")
```

Reply:
✔️  Good catch, thank you.